pkg/status/combinedstatus-resolver.go

```diff
@@ -0,0 +1,520 @@
+	} else {
+		delete(c.statusCollectorNameToSpec, statusCollector.Name)
+	}
+
+	return combinedStatusIdentifiersToQueue
+}
+
+// ResolutionExists returns true if a combinedstatus resolution exists for
+// the given binding name and workload object identifier.
+func (c *combinedStatusResolver) ResolutionExists(bindingName string, objectIdentifier util.ObjectIdentifier) bool {
+	c.RLock()
+	defer c.RUnlock()
+
+	if resolutions, exists := c.bindingNameToResolutions[bindingName]; exists {
+		_, exists = resolutions[objectIdentifier]
+		return exists
+	}
+
+	return false
+}
+
+// fetchMissingStatusCollectorSpecs fetches the missing statuscollector specs
+// from the given lister and updates the cache.
+// The method is expected to be called with the write lock held.
+func (c *combinedStatusResolver) fetchMissingStatusCollectorSpecsLocked(statusCollectorLister controllisters.StatusCollectorLister,
+	statusCollectorNames []string) {
+	for _, statusCollectorName := range statusCollectorNames {
+		if _, exists := c.statusCollectorNameToSpec[statusCollectorName]; exists {
+			continue // this method is not responsible for keeping the cache up-to-date
+		}
+
+		statusCollector, err := statusCollectorLister.Get(statusCollectorName)
+		if err != nil {
+			// fetch error should not disturb the flow.
+			// a missing spec will be reconciled when the status collector is created/updated.
+			runtime2.HandleError(fmt.Errorf("failed to get statuscollector %s: %w", statusCollectorName, err))
+			return
+		}
+
+		c.statusCollectorNameToSpec[statusCollectorName] = &statusCollector.Spec // readonly
+	}
+}
+
+// evaluateWorkStatusesPerBindingReadLocked evaluates workstatuses associated with the given
+// workload identifiers and destinations. Workstatuses that are not associated
+// with the two are skipped.
+// The returned set contains the identifiers of combinedstatus objects that
+// should be queued for syncing.
+// The method is expected to be called with the read lock held.
+func (c *combinedStatusResolver) evaluateWorkStatusesPerBindingReadLocked(bindingName string,
+	workloadObjIdentifiersToEvaluate sets.Set[util.ObjectIdentifier], destinations sets.Set[string],
+	workStatusLister cache.GenericLister) sets.Set[util.ObjectIdentifier] {
+	workloadObjectToError := make(map[util.ObjectIdentifier]error) // TODO: use
+	combinedStatusesToQueue := sets.Set[util.ObjectIdentifier]{}
+
+	for workloadObjIdentifier := range workloadObjIdentifiersToEvaluate {
+		// fetch workstatuses
+		workStatuses, err := getWorkStatusesOfObject(workStatusLister, workloadObjIdentifier)
+		if err != nil {
+			runtime2.HandleError(fmt.Errorf("failed to get workstatuses for workload object %v: %w",
+				workloadObjIdentifier, err))
+			continue
+		}
+
+		// evaluate workstatuses
+		for _, workStatus := range workStatuses {
+			// skip workstatuses that are not associated with the binding's destinations
+			if destinations.Has(workStatus.wecName) {
+				continue
+			}
```

Comment:
Rather than fetch and convert all of the WorkStatus objects and then throw away the ones wanted, this code could iterate over `destinations` and for each one just fetch and convert the one WorkStatus object for that destination.

This is an efficiency issue, not correctness. Can be addressed in follow-on.

Reply:
thanks